File: cis4020/lectures/linkedlists/linkedlist-part-3-cis4020/linkedlist-cpp/include/LinkedList.cpp
#ifndef LINKED_LIST_TPP
#define LINKED_LIST_TPP

template <typename T>
LinkedList<T>::LinkedList() : head(nullptr) {}

template <typename T>
LinkedList<T>::~LinkedList() {
    destroy();
}

template <typename T>
bool LinkedList<T>::isEmpty() const {
    return head == nullptr;
}

template <typename T>
void LinkedList<T>::addFirst(const T& value) {
    Node<T>* newNode = new Node<T>(value);
    newNode->next = head;
    head = newNode;
}

template <typename T>
void LinkedList<T>::addLast(const T& value) {
    Node<T>* newNode = new Node<T>(value);
    if (isEmpty()) {
        head = newNode;
        return;
    }
    Node<T>* current = head;
    while (current->next != nullptr) {
        current = current->next;
    }
    current->next = newNode;
}

template <typename T>
T LinkedList<T>::retrieve(int position) const {
    if (position < 0) {
        throw std::out_of_range("position must be non-negative");
    }
    Node<T>* current = head;
    int index = 0;
    while (current != nullptr) {
        if (index == position) {
            return current->data;
        }
        current = current->next;
        index++;
    }
    throw std::out_of_range("position is past the end of the list");
}

template <typename T>
bool LinkedList<T>::remove(const T& value) {
    if (isEmpty()) {
        return false;
    }
    if (head->data == value) {
        Node<T>* toDelete = head;
        head = head->next;
        delete toDelete;
        return true;
    }
    Node<T>* current = head;
    while (current->next != nullptr && !(current->next->data == value)) {
        current = current->next;
    }
    if (current->next == nullptr) {
        return false;
    }
    Node<T>* toDelete = current->next;
    current->next = toDelete->next;
    delete toDelete;
    return true;
}

template <typename T>
void LinkedList<T>::display() const {
    Node<T>* current = head;
    while (current != nullptr) {
        std::cout << current->data << " ";
        current = current->next;
    }
    std::cout << std::endl;
}

// Walks the list and prints each node's own address, its data, and the
// address it points to next, so you can see the chain of pointers the same
// way the Java walk() method shows the chain of object identities.
template <typename T>
void LinkedList<T>::walk() const {
    std::cout << std::left;
    std::cout << "head: " << ptrLabel(head) << std::endl;
    Node<T>* current = head;
    while (current != nullptr) {
        std::ostringstream dataStream;
        dataStream << current->data;
        std::cout << std::setw(16) << ptrLabel(current)
                  << std::setw(14) << dataStream.str()
                  << ptrLabel(current->next) << std::endl;
        current = current->next;
    }
    std::cout << std::endl;
}

// needed by the walk method
template <typename T>
std::string LinkedList<T>::ptrLabel(const Node<T>* p) {
    if (p == nullptr) {
        return "nullptr";
    }
    std::ostringstream oss;
    oss << p;
    return oss.str();
}

template <typename T>
void LinkedList<T>::destroy() {
    while (head != nullptr) {
        Node<T>* temp = head;
        head = head->next;
        delete temp;
    }
}

#endif

File: cis4020/lectures/linkedlists/linkedlist-part-3-cis4020/linkedlist-java-class-edited/src/linkedlist/LinkedList.java
package linkedlist;

public class LinkedList<T> {
    
    private Node<T> head;

    public boolean isEmpty() {
        return head == null;
    }

    // Adds to the front. Cheapest place to insert, since there is no tail reference.
    public void addFirst(T value) {
        Node<T> newNode = new Node<>(value);
        newNode.setNext(head);
        head = newNode;
    }
    
    public void addLast(T value) {
        Node<T> newNode = new Node<>(value);
        if (isEmpty()) {
            head = newNode;
            return;
        }
        Node<T> current = head;
        while (current.getNext() != null) {
            current = current.getNext();
        }
        current.setNext(newNode);
    }    

    // 0-based position. Throws if the list is empty or the position does not exist.
    public T retrieve(int position) {
        if (position < 0) {
            throw new IndexOutOfBoundsException("position must be non-negative");
        }
        Node<T> current = head;
        int index = 0;
        while (current != null) {
            if (index == position) {
                return current.getData();
            }
            current = current.getNext();
            index++;
        }
        throw new IndexOutOfBoundsException("position is past the end of the list");
    }

    // Removes the first node whose data equals value. Relies on T's equals().
    public boolean remove(T value) {
        if (isEmpty()) {
            return false;
        }
        if (head.getData().equals(value)) {
            head = head.getNext();
            return true;
        }
        Node<T> current = head;
        while (current.getNext() != null && !current.getNext().getData().equals(value)) {
            current = current.getNext();
        }
        if (current.getNext() == null) {
            return false;   // value was not found
        }
        current.setNext(current.getNext().getNext());
        return true;
    }

    public void display() {
        Node<T> current = head;
        while (current != null) {
            System.out.print(current.getData() + " ");
            current = current.getNext();
        }
        System.out.println();
    }
    
    public void reverse() {
    	reversedisplay(head);
    }

    private void reversedisplay(Node<T> node) {
    	if (node == null) {
    		return;
    	}
    	reversedisplay(node.getNext());
    	System.out.println(node.getData() + " ");
    	
    }
    
    // Walks the list and prints each node's own identity, its data, and the
    // identity of the node it points to next.
    public void walk() {
        Node<T> current = head;
        System.out.println("Head:" + head);
        while (current != null) {
            System.out.printf("%-24s %-11s %s\n", current, current.getData(), current.getNext());
            current = current.getNext();
        }
        System.out.println();
    }
    
    // Empties the list. Nothing references the old nodes afterward, so the
    // garbage collector reclaims them; there is no delete to call.
    public void destroy() {
        head = null;
    }
}
